src/UserSpaceInstrumentation/RegisterState.cpp

```diff
@@ -0,0 +1,140 @@
+
+namespace {
+
+// Return the size of the XSave area on this cpu.
+[[nodiscard]] ErrorMessageOr<size_t> GetXSaveAreaSize() {
+  uint32_t eax = 0;
+  uint32_t ebx = 0;
+  uint32_t ecx = 0;
+  uint32_t edx = 0;
+  if (!__get_cpuid(1, &eax, &ebx, &ecx, &edx) || !(ecx & bit_XSAVE)) {
+    return ErrorMessage("XSAVE is not supported by Cpu.");
+  }
+  if (!__get_cpuid_count(0x0d, 0x00, &eax, &ebx, &ecx, &edx)) {
+    return ErrorMessage("XSAVE is not supported by Cpu.");
+  }
+  return static_cast<size_t>(ecx);
+}
+
+// Return offset of the YMMx registers inside the extended section of the XSave area.
+[[nodiscard]] ErrorMessageOr<size_t> GetAvxOffset() {
+  uint32_t eax = 0;
+  uint32_t ebx = 0;
+  uint32_t ecx = 0;
+  uint32_t edx = 0;
+  if (!__get_cpuid(0x01, &eax, &ebx, &ecx, &edx) || !(ecx & bit_AVX)) {
+    return ErrorMessage("AVX is not supported by Cpu.");
+  }
+  if (!__get_cpuid_count(0x0d, 0x02, &eax, &ebx, &ecx, &edx)) {
+    return ErrorMessage("AVX offset query failed.");
+  }
+  return static_cast<size_t>(ebx);
+}
+
+}  // namespace
+
+bool RegisterState::Hasx87DataStored() {
+  return (static_cast<uint64_t>(GetXSaveHeader()->xstate_bv) &
+          static_cast<uint64_t>(XSaveHeader::StateComponents::kX87)) != 0;
+}
+
+bool RegisterState::HasSseDataStored() {
+  return (static_cast<uint64_t>(GetXSaveHeader()->xstate_bv) &
+          static_cast<uint64_t>(XSaveHeader::StateComponents::kSse)) != 0;
+}
+
+bool RegisterState::HasAvxDataStored() {
+  return (static_cast<uint64_t>(GetXSaveHeader()->xstate_bv) &
+          static_cast<uint64_t>(XSaveHeader::StateComponents::kAvx)) != 0;
+}
+
+[[nodiscard]] ErrorMessageOr<void> RegisterState::BackupRegisters(pid_t tid) {
+  tid_ = tid;
+
+  iovec iov;
+  iov.iov_base = &general_purpose_registers_;
+  iov.iov_len = sizeof(GeneralPurposeRegisters);
+  auto result = ptrace(PTRACE_GETREGSET, tid, NT_PRSTATUS, &iov);
+  if (result == -1) {
+    return ErrorMessage(absl::StrFormat("PTRACE_GETREGS, NT_PRSTATUS failed with errno: %d: \"%s\"",
+                                        errno, SafeStrerror(errno)));
+  }
+
+  if (iov.iov_len == sizeof(GeneralPurposeRegisters32)) {
+    bitness_ = Bitness::k32Bit;
+  } else if (iov.iov_len == sizeof(GeneralPurposeRegisters64)) {
+    bitness_ = Bitness::k64Bit;
+  } else {
+    CHECK(false);
+  }
+
+  auto xsave_area_size = GetXSaveAreaSize();
+  if (xsave_area_size.has_error()) {
+    return xsave_area_size.error();
+  }
+  xsave_area_.resize(xsave_area_size.value());
+
+  iov.iov_len = xsave_area_.size();
+  iov.iov_base = xsave_area_.data();
+  result = ptrace(PTRACE_GETREGSET, tid, NT_X86_XSTATE, &iov);
+  if (result == -1) {
+    return ErrorMessage(absl::StrFormat(
+        "PTRACE_GETREGS, NT_X86_XSTATE failed with errno: %d: \"%s\"", errno, SafeStrerror(errno)));
+  }
+
+  auto avx_offset = GetAvxOffset();
+  if (!avx_offset.has_error()) {
+    avx_offset_ = avx_offset.value();
+  } else {
+    avx_offset_ = 0;
```

Comment:
Is `0` the real value in case of error or is this an assumption? In other words, why isn't an `ErrorMessage` returned here instead?